File: src/test_generator.py
"""
Core test generation engine using Deepseek LLM.
"""
from typing import Optional

from src.config import Config
from src.api_client import DeepseekAPIClient
from src.prompt_engineer import build_test_generation_prompt
from src.utils import extract_code_blocks, validate_python_code, write_file


class TestGenerator:
    """Generate unit tests for given source code using an LLM."""

    def __init__(self, config: Config, client: Optional[DeepseekAPIClient] = None):
        self.config = config
        self.client = client or DeepseekAPIClient(config)

    def generate_tests(
        self,
        code: str,
        specs: Optional[str] = None,
        framework: str = "pytest",
    ) -> str:
        """
        Generate unit tests for the provided source code.

        Args:
            code: Source code to test
            specs: Optional specifications or behavior
            framework: Test framework to use

        Returns:
            Python test module content as string
        """
        prompt = build_test_generation_prompt(code=code, specs=specs, framework=framework)
        raw = self.client.generate_text(prompt)

        # Prefer python code block if present
        blocks = extract_code_blocks(raw, language="python")
        test_code = blocks[0] if blocks else raw

        # Validate and fallback to simple wrapper if invalid
        if not validate_python_code(test_code):
            test_code = f"# Generated tests (raw)\n{test_code}"

        return test_code

    def generate_and_save(
        self,
        code: str,
        output_path: str,
        specs: Optional[str] = None,
        framework: str = "pytest",
        overwrite: bool = False,
    ) -> str:
        """
        Generate tests and save to a file.

        Returns the saved path.
        """
        tests = self.generate_tests(code=code, specs=specs, framework=framework)
        write_file(output_path, tests, overwrite=overwrite)
        return output_path

File: backend/src/services/testGenerator.js
import llmClient from './llmClient.js';
import { buildTestGenerationPrompt, extractCodeBlocks } from '../utils/prompts.js';

/**
 * Test Generator Service
 * Core logic for generating unit tests from source code
 */
class TestGenerator {
    /**
     * Generate unit tests for provided source code
     * @param {object} params - Generation parameters
     * @param {string} params.code - Source code to test
     * @param {string} [params.specs] - Optional specifications
     * @param {string} [params.framework] - Test framework (default: pytest)
     * @param {string} [params.language] - Programming language (default: python)
     * @param {string} [params.provider] - LLM provider (gemini/deepseek)
     * @param {string} [params.userApiKey] - User's own API key
     * @returns {Promise<object>} Generated tests and metadata
     */
    async generateTests({ code, specs = '', framework = 'pytest', language = 'python', provider, userApiKey }) {
        if (!code || code.trim().length === 0) {
            throw new Error('Source code is required');
        }

        console.log(`📝 Generating ${framework} tests for ${language} code using ${provider || 'default'} provider...`);

        // Build the prompt
        const prompt = buildTestGenerationPrompt({
            code,
            specs,
            framework,
            language,
        });

        // Call LLM API with optional user key
        const startTime = Date.now();
        const rawResponse = await llmClient.generateText(prompt, {
            provider,
            userApiKey,
        });
        const generationTime = Date.now() - startTime;

        // Extract code from response
        const codeBlocks = extractCodeBlocks(rawResponse, language);
        const testCode = codeBlocks.length > 0 ? codeBlocks[0] : rawResponse;

        // Validate the generated code (basic check)
        const isValid = this._validateCode(testCode, language);

        console.log(`✅ Tests generated in ${generationTime}ms`);

        return {
            generatedTests: testCode,
            rawResponse,
            framework,
            language,
            generationTime,
            isValid,
            timestamp: new Date().toISOString(),
        };
    }

    /**
     * Basic code validation
     * @private
     */
    _validateCode(code, language) {
        if (!code || code.trim().length === 0) {
            return false;
        }

        // Basic checks for common issues
        if (language === 'python') {
            // Check for Python test patterns
            return code.includes('def test_') ||
                code.includes('class Test') ||
                code.includes('assert ') ||
                code.includes('unittest');
        }

        if (language === 'javascript' || language === 'typescript') {
            // Check for JS test patterns
            return code.includes('describe(') ||
                code.includes('it(') ||
                code.includes('test(') ||
                code.includes('expect(');
        }

        return true; // Default to valid for unknown languages
    }
}

// Singleton instance
const testGenerator = new TestGenerator();

export default testGenerator;
export { TestGenerator };
